More time functions

netsettler opened the pull request:
**Description of Change**

In `misc_utils`:

* Add `as_ref_datetime` to convert times to the reference timezone (US/Eastern by default).
* Add `as_utc_datetime` to convert times to UTC.
* Rename `HMS_TZ` to `REF_TZ`, but keep `HMS_TZ` as a synonym for compatibility for now.
* Rename `hms_now` to `ref_now`, but again keep `hms_now` as a synonym for compatibility for now.

**Rationale**

If we deploy at other locations, it may not be HMS that is relevant, so we could be at some place with another timezone.

willronchetti commented on dcicutils/misc_utils.py:
```diff
@@ -999,3 +1029,8 @@ def delete(self, file):
         return value
+
+
+# Deprecated names, still supported for a while.
+HMS_TZ = REF_TZ
+hms_now = ref_now
```
Is it convention to do this at the bottom of the file? I would think it makes sense to put these next to the respective definition ie: by `REF_TZ`.

netsettler replied:
It's _a_ convention, but I'm not wedded to it if it bugs you. The idea is to say it's an after-the-fact thing to not mention these names anywhere in the basic file and only at the end. Putting them here puts them out of scope to refer to at top-level. Python makes it hard to put them out of scope from function definitions without putting them in another file, which wouldn't accomplish the purpose.

netsettler replied:
I didn't get an answer on this, but I'm inclined to think my rationale is reasonable so I think I'll leave this as-is.

willronchetti commented on dcicutils/misc_utils.py:
```diff
@@ -464,6 +464,33 @@ def as_datetime(dt, tz=None):
+    If the time is already a datetime, no parsing occurs, but the time is still adjusted to present as it would in the
+    reference timeszone.
+    """
+    dt = as_datetime(dt, tz=REF_TZ)
+    hms_dt = dt.astimezone(REF_TZ)
```
Isn't it possible for `as_datetime` to return None, hence throwing `AttributeError` below? Or do you expect that will never happen with this function? Same question in the next function.

netsettler replied:
Good catch. It may not come up but should behave predictably there.

netsettler replied:
Fixed. Added unit testing.

willronchetti reviewed (approved):
Approving with some small comments

netsettler commented:
I ran tests on my machine since this is all clogged up on Travis. It tests out fine and I'm going to merge it.
```
(utl9env) kentpitman@Kents-MacBook-Pro ~/py/utils9$ make test
pytest -vv
===================================================================================== test session starts =====================================================================================
platform darwin -- Python 3.6.9, pytest-6.0.1, py-1.9.0, pluggy-0.13.1 -- /Users/kentpitman/py/utils9/utl9env/bin/python
cachedir: .pytest_cache
rootdir: /Users/kentpitman/py/utils9, configfile: pytest.ini
plugins: flaky-3.7.0, mock-3.3.1, cov-2.10.1
collected 299 items                                                                                                                                                                           

test/test_beanstalk_utils.py::test_get_beanstalk_real_url_blue_data PASSED [  0%]
... many passing tests elided ...
test/test_s3_utils.py::test_unzip_s3_to_s3_store_results PASSED [100%]
===Flaky Test Report===

test_unified_authentication passed 1 out of the required 1 times. Success!
test_get_authentication_with_server passed 1 out of the required 1 times. Success!
test_authorized_request_integrated passed 1 out of the required 1 times. Success!
test_get_metadata passed 1 out of the required 1 times. Success!
test_patch_metadata passed 1 out of the required 1 times. Success!
test_post_delete_purge_links_metadata passed 1 out of the required 1 times. Success!
test_upsert_metadata passed 1 out of the required 1 times. Success!
test_get_search_generator passed 1 out of the required 1 times. Success!
test_get_es_metadata passed 1 out of the required 1 times. Success!
test_get_es_search_generator passed 1 out of the required 1 times. Success!
test_get_health_page passed 1 out of the required 1 times. Success!
test_get_schema_names passed 1 out of the required 1 times. Success!
test_expand_es_metadata passed 1 out of the required 1 times. Success!
test_expand_es_metadata_frame_object_embedded passed 1 out of the required 1 times. Success!
test_expand_es_metadata_add_wfrs passed 1 out of the required 1 times. Success!
test_expand_es_metadata_complain_wrong_frame passed 1 out of the required 1 times. Success!
test_expand_es_metadata_ignore_fields passed 1 out of the required 1 times. Success!
test_delete_field passed 1 out of the required 1 times. Success!
test_dump_results_to_json passed 1 out of the required 1 times. Success!

===End Flaky Test Report===

========== 298 passed, 1 skipped in 471.45s (0:07:51) ==========
(utl9env) kentpitman@Kents-MacBook-Pro ~/py/utils9$ git status
On branch kmp_more_time_functions
Your branch is up to date with 'origin/kmp_more_time_functions'.

nothing to commit, working tree clean
```

netsettler commented:
I made a couple more last minute changes to regularize the expectation for likely consumers of this code:
* Changed `as_datetime` to raise an error on bad data, so it matches what `as_utc_datetime` and `as_ref_datetime` do. It was not documented what would happen in this case, and I'm going to treat it as a bug that it didn't do so before. Only `brig` probably cares, and the code that cares is not released yet.
* Changed `in_datetime_interval` to use `as_ref_datetime` on all of its arguments (except if `start` or `end` is `None`). This will simplify the calling sequence in `4dn-status` in `brig`.
I reran tests and got the same test output on my local machine as pasted above, so I won't clutter this with another copy.

netsettler commented:
@willronchetti Some of those final changes are judgment calls as to whether they were responsive to code review or something new, but in fixing your question about the `None` being returned I just decided that was too error-prone and it was better for programs to request that behavior if they want it. If you have problems, I can submit another PR but I want to just get this merged so we're not wasting time this week on it, especially with Travis a mess. I tested this on my machine and we should be good to go.
